### PR TITLE
get top x quiz results

### PR DESCRIPTION
**Pull request by Adam3004**

### CODE REVIEW

File: src/main/resources/api.yaml [hunk 1]
```diff
@@ -438,6 +438,30 @@ paths:
...
           description: Quiz not found
+  /quizzes/{quizId}/results/{numberOfTopResults}:
+    get:
+      summary: Get top x results of a specific quiz
+      operationId: getTopResultsByQuizId
+      tags:
+        - quizzes
+      parameters:
+        - '$ref': '#/components/parameters/quizId'
+        - '$ref': '#/components/parameters/numberOfTopResults'
```
Comment (krisx303) [inline src/main/resources/api.yaml:448]: Why not query param?

File: src/main/resources/api.yaml [hunk 2]
```diff
@@ -438,6 +438,30 @@ paths:
           description: User do not have permissions to get quiz results
         '404':
           description: Quiz not found
+  /quizzes/{quizId}/results/{numberOfTopResults}:
```
Comment (krisx303) [inline src/main/resources/api.yaml]: I would change that to ` /quizzes/{quizId}/results`

File: src/main/java/com/brightpath/learnify/domain/quiz/Quiz.java [hunk 3]
```diff
@@ -16,7 +17,6 @@ public record Quiz(UUID id,
                    QuizSimpleResult lastScore,
```
Comment (krisx303) [inline src/main/java/com/brightpath/learnify/domain/quiz/Quiz.java:17]: Hmm, should this class still contain lastScore and bestScore?

File: src/main/java/com/brightpath/learnify/persistance/quiz/QuizRepository.java [hunk 4]
```diff
@@ -15,6 +15,6 @@ public interface QuizRepository extends JpaRepository<QuizEntity, UUID> {
             """)
     List<QuizEntity> searchQuizzes(UUID workspaceId);
 
-    @Query("SELECT e FROM QuizEntity e ORDER BY COALESCE(e.lastTryDate, e.createdAt) DESC LIMIT 4")
-    List<QuizEntity> findTop4RecentQuizzes();
+    @Query("SELECT e FROM QuizEntity e ORDER BY COALESCE((select r.lastTryDate FROM e.quizResults as r WHERE r.userId=userId), e.createdAt) DESC LIMIT 4")
+    List<QuizEntity> findTop4RecentQuizzes(String userId);
```
Comment (krisx303) [inline src/main/java/com/brightpath/learnify/persistance/quiz/QuizRepository.java]: nice

File: src/main/java/com/brightpath/learnify/domain/quiz/QuizService.java [hunk 5]
```diff
@@ -168,4 +178,20 @@ public void deleteQuiz(UUID quizId) {
         bindingService.removeBindingForQuiz(quizId);
         quizRepository.deleteById(quizId);
     }
+
+    public List<QuizUserResult> getTopQuizResults(UUID quizId, int numberOfExpectedResults) {
+        QuizEntity quiz = quizRepository.getReferenceById(quizId);
+        List<QuizResultsEntity> topResults = quiz.getQuizResults().stream()
```
Comment (krisx303) [inline src/main/java/com/brightpath/learnify/domain/quiz/QuizService.java:186]: But why do we have to do all of that at the service level and not database?

**Review by krisx303 — APPROVED** [review]

File: src/main/resources/api.yaml (hunk 2 above)
Comment (krisx303) [reply src/main/resources/api.yaml]: I approved, but consider this change. Not sure if this is some other RESTful convention but I really don't like that xD

File: src/main/java/com/brightpath/learnify/domain/quiz/Quiz.java (hunk 3 above)
Comment (Adam3004) [reply src/main/java/com/brightpath/learnify/domain/quiz/Quiz.java:17]: yes, it is set for requesting user so it's ok. Moreover, we use those values on frontend

**Pull request merged**